src/clients/kol.ts

```diff
@@ -87,95 +114,239 @@ function sanitiseBlueText(blueText: string | undefined): string {
-            withCredentials: true,
-            headers: {
-              cookie: sessionCookies,
-            },
-            params: {
-              what: "status",
-              for: "OAF Discord bot for Kingdom of Loathing",
-            },
-          });
-          this._credentials = {
-            fetched: new Date().getTime(),
-            sessionCookies: sessionCookies,
-            pwdhash: apiResponse.data.pwd,
-          };
-        } else {
-          console.log("Blocked fetching new credentials");
-          console.log(
-            `${60000 + this._credentials.fetched - new Date().getTime()} milliseconds to new login`
-          );
-        }
-      } catch (error) {
-        console.log(error);
+        const loginResponse = await axios("https://www.kingdomofloathing.com/login.php", {
+          method: "POST",
+          data: this._loginParameters,
+          maxRedirects: 0,
+          validateStatus: (status) => status === 302,
+        });
+        const sessionCookies = (loginResponse.headers["set-cookie"] || [])
+          .map((cookie: string) => cookie.split(";")[0])
+          .join("; ");
+        const apiResponse = await axios("https://www.kingdomofloathing.com/api.php", {
+          withCredentials: true,
+          headers: {
+            cookie: sessionCookies,
+          },
+          params: {
+            what: "status",
+            for: `${this._loginParameters.get("loginname")} Chatbot`,
+          },
+        });
+        this._credentials = {
+          sessionCookies: sessionCookies,
+          pwdhash: apiResponse.data.pwd,
+        };
+        return true;
+      } catch {
+        console.log("Login failed. Checking if it's because of rollover.");
+        await this.rolloverCheck();
+        return false;
       }
     });
   }
 
-  private async makeCredentialedRequest(url: string, parameters: object) {
+  private chatBotStarted = false;
+
+  async startChatBot() {
+    if (this.chatBotStarted) return;
+    this.loopChatBot();
+    this.chatBotStarted = true;
+  }
+
+  private async loopChatBot() {
+    await Promise.all([this.checkWhispers(), this.checkKmails()]);
+    await wait(3000);
+    await this.loopChatBot();
+  }
+
+  private lastFetchedWhispers = "0";
+
+  async checkWhispers() {
+    const newChatMessagesResponse = (await this.visitUrl("newchatmessages.php", {
+      j: 1,
+      lasttime: this.lastFetchedWhispers,
+    })) as { last: string; msgs: KoLChatMessage[] };
+
+    if (!newChatMessagesResponse || typeof newChatMessagesResponse !== "object") return;
+
+    this.lastFetchedWhispers = newChatMessagesResponse["last"];
+
+    newChatMessagesResponse["msgs"]
+      .filter((msg) => msg["type"] === "private" && msg.who && msg.msg)
+      .map((msg) => ({
+        who: {
+          id: Number(msg.who!.id),
+          name: msg.who!.name,
+        },
+        msg: msg.msg!,
+        time: new Date(Number(msg.time) * 1000),
+      }))
```

Comment:
```suggestion
      .filter((msg): msg is { type: "private", who: string, msg: string. time?: unknown} => msg["type"] === "private" && msg.who && msg.msg)
      .map((msg) => ({
        who: {
          id: Number(msg.who.id),
          name: msg.who.name,
        },
        msg: msg.msg,
        time: new Date(Number(msg.time) * 1000),
      }))
```
feels more elegant to add a typeguard to the filter than to use nonnull assertions